### ISSUE TITLE
macOS/linux Support?

### ISSUE THREAD

**Issue by ThomasAlban** [TeamWheelWizard/WheelWizard#3]
Would be great!
Not sure how possible this is in the current state of the project. I can see throughout your code you have `using System.Windows` which isn't a good sign!

**Comment by rehtlaw** [comment]
As far as I can tell, it's a WPF app, so the UI would have to be redone in something cross-platform (Avalonia for example).
Good to know I'm not the only one wishing it wasn't Windows-only.

**Comment by soccerboys2008** [comment]
hmm, after i saw this issue i decided to see if i can open it using wine on linux

well, sort of:
[Screencast from 07-09-2024 05:02:27 PM.webm](https://github.com/patchzyy/WheelWizard/assets/29149297/04d51a3b-d4ad-49dd-8bf7-524e513f08aa)

so a couple things

1. can we change the title to "macOS and Linux Support"
2. anything you guys want me to try after forcing it with wine (not sure exactly what this app is capable of, just immediately came here after hearing troy shout it out on his stream)

**Comment by soccerboys2008** [comment]
as a note, whenever you click a tab or something in the app, i have to move the app to a different workspace in order to get it to render (as in the clip above), but it seems to be able to interact with wines interface

doubt it can launch dolphin on linux though (at least it won't be able to launch the flatpak version i have installed, which is the one everyone uses (cause nobody wants to compile it from source), if devs are curious we can open dolphin with a command using the terminal (in my case (and for all flatpak users) its "flatpak run org.DolphinEmu.dolphin-emu", but for those who compile from source, they can just add the binary to their path)

**Comment by sonalder-darlene** [comment]
I would also like having a linux version of the game as I run ChimeraOS and SteamOS with dolphin. This would make things easier ^^

**Comment by patchzyy** [comment]
yes, in order to release it for mac and linux we would have to port the program over using something like Avalonia. We will first continue to add all features to make the program fully functional, and after that we will explore the options of porting it to Avalonia for linux and mac OS. In the meantime you may manually install Retro Rewind [here](https://www.youtube.com/watch?v=qH4ou21r8ic&t=2s)

**Comment by soccerboys2008** [comment]
@sonalder-darlene have you been able to get retro rewind to work on linux using the manual method

**Comment by rehtlaw** [comment]
The manual install works just fine on any OS that Dolphin supports, it's just a matter of getting this updater to be compatible with MacOS/Linux as well.

**Comment by soccerboys2008** [comment]
Yeah, I think I found the root cause

When I compiled dolphin from source it worked

But when I used the flatpak version of dolphin I got a empty message box with 2 options "start without saving" and return to retro rewind channel"

The difference between the flatpak version and my compile is only 2 thing

1. The version (2407-39 (for the one I compiled), and 2407 (for flatpak))
2. Flatpak version is sandboxed, amongst other things, permissions, etc.


So either dolphin devs already fixed whatever caused it, or there's an issue with the flatpak

Considering the tiny amount of dev versions between the 2

Its probably the flatpak

![Screenshot from 2024-07-10 17-09-03](https://github.com/patchzyy/WheelWizard/assets/29149297/31fe0d73-cfc0-42c7-8daa-b13a2117ec41)

**Comment by soccerboys2008** [comment]
And if it was something fixed by dolphin devs, then the flatpak just needs to be updated

**Comment by sonalder-darlene** [comment]
> @sonalder-darlene have you been able to get retro rewind to work on linux using the manual method

Yes it wasn't hard, I also managed to add it as a non steam game to the steam launcher.

**Comment by rehtlaw** [comment]
> Yeah, I think I found the root cause
> 
> When I compiled dolphin from source it worked
> 
> But when I used the flatpak version of dolphin I got a empty message box with 2 options "start without saving" and return to retro rewind channel"
> 
> The difference between the flatpak version and my compile is only 2 thing
> 
> 1. The version (2407-39 (for the one I compiled), and 2407 (for flatpak))
> 2. Flatpak version is sandboxed, amongst other things, permissions, etc.
> 
> So either dolphin devs already fixed whatever caused it, or there's an issue with the flatpak
> 
> Considering the tiny amount of dev versions between the 2
> 
> Its probably the flatpak
> 

How did you install it? Adding MKWii to my Dolphin flatpak's library and putting the RetroRewind6 and riivolution folders inside of `$HOME/.var/app/org.DolphinEmu.dolphin-emu/data/dolphin-emu/Load/Riivolution/` and starting up the game with the patches enabled gave the exact same result as my regular Dolphin version.

**Comment by soccerboys2008** [comment]
Load folder location was the difference, I had set the load folder to "$HOME/Wii/Dolphin Load Folder" to be wwith the rest of the Wii stuff for easier access

Moving it back to its original location fixes the issue

So it's a flatpak permissions issue / sandbox issue (unsure how restrictive the sandbox is)

**Comment by rehtlaw** [comment]
Are you sure being able to load riivolution mods from outside of `$XDG_DATA_HOME` is intended?

**Comment by ppebb** [comment]
I wouldn't mind redoing the GUI in something like GTK through Gir Core for mac and linux. Is a PR for something like that welcome?

**Comment by GabrieleAncillai** [comment]
Hey there! Another linux user here, I'm really interested in porting this app to also support the SteamDeck, I do 98% of all my gaming in that thing and I'm using RetroDeck, so... supporting  RetroDeck and EmuDeck Dolphin installation would be really great as well.

Also I use fedora on my main rig with Dolphin installed via flatpak.

My recommendation about packaging and distributing this app would be to do it via flatpak/flathub for peace of mind, I can help with that too.

**Comment by rehtlaw** [comment]
Yeah I think we're just waiting on word from @patchzyy on how they wanna move forward, since we have people that are willing to help out with porting it to something crossplatform.

**Comment by soccerboys2008** [comment]
@GabrieleAncillai there is a slight pitfall on the flatpak version of dolphin that specifically affects this riivolution mod atm (its referenced above), basically if a user sets the dolphin load folder to a location outside the sandbox, it causes a weird issue with the mod, (as dolphin doesn't have write access)

this doesn't happen with other mods as they don't keep save data in LoadFolder/Riivolution/riivolution

who knows how many users change there load folders location, but if they do it can happen

**Comment by soccerboys2008** [comment]
I think a good solution for those trying to make a Linux port is to add a banner warning at the top

Maybe something like "If your using the flatpak version of dolphin, do not change the location of the dolphin loads folder from its default"

**Comment by blunelsongamer** [comment]
The best way to open Wheel Wizard is bottles.

Bottles is a flatpak alternative to Wine with prefixes and latest updates without too much of a hostile. A simple way ofc is to created a application bottle, run executable, click on wheel wizard, and bam.

Now I'm not sure if it's fully stable, but I'm sure it actually works!

**Comment by soccerboys2008** [comment]
not so sure that's a good idea, with dolphin having a native version on linux (whell wizard needs a path to dolphin specified, and since it would be in a wine environment it would open dolphin in the same environment)

currently its best to manually setup dolphin on linux (don't change default location for dolphin loads folder, if using flatpak) and mac

this way your running natively, rather then having calls go through a translation layer

**Comment by rehtlaw** [comment]
Yeah, if all you care about is seeing the WFC lobbies, then running it through wine is fine, but I wouldn't want to do that if you actually want to update the mod and run it through WheelWizard, hence why people are asking for it to be ported.

**Comment by GabrieleAncillai** [comment]
Hi there, brief update:
I tried compiling  this project for linux and I'm having the following issue:
```
usr/lib64/dotnet/sdk/8.0.108/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.DefaultItems.targets(141,5): error 
NETSDK1136: The target platform must be set to Windows (usually by including '-windows' in the TargetFramework 
property) when using Windows Forms or WPF, or referencing projects or packages that do so. [/home/muffin/Documents/
Projects/WheelWizard/WheelWizard/WheelWizard.csproj::TargetFramework=netcoreapp8.0
```
Here's more info about this specific case [Error NETSDK1136](https://stackoverflow.com/questions/74358983/error-netsdk1136-the-target-platform-must-be-set-to-windows)

I was going to work on a native version for this app for linux, unfortunately for time availability reasons I'll not be working on fixing this.  If there's someone who would  then I could give it a second chance, but tbh I'd prefer working on re-doing the whole app on another friendlier language

**Comment by ppebb** [comment]
There's no issue with continuing to use dotnet. The issues present are because of the project's use of WPF, which is windows exclusive. Redoing the GUI to use something cross-platform would be preferable to redoing the entire application.

**Comment by GabrieleAncillai** [comment]
> The issues present are because of the project's use of WPF, which is windows exclusive.

Yeap, that's exactly the issue, I just proposed redoing the app bc I don't like .net but that's my personal preference. Fixing the issue would imply me taking too much time bc I don't really know what I'm doing and my main goal is packaging the app, not fixing compatibility. If someone fixes it then I would be able to stark working in bundles

Best regards y'all :)

**Comment by ppebb** [comment]
I worked on one for a bit and made progress, but I'm incredibly busy for now and did not finish. If no one (e.g. the actual maintainers) makes something xplat by the time I'm free I'll look at a restructure again.

**Comment by patchzyy** [comment]
We have started working on an official avalonia port of wheel wizard

**Comment by FerLuisxd** [comment]
Maybe a stretch but I would love to see an android version too!

**Comment by WantToBeeMe** [comment]
Just a little update. We have been working on it for a while now. 2 devs with minimal time so it took a while.
However, The multiplatform version is almost done and will soon be ready for the first tests.

**Comment by patchzyy** [comment]
it has been done...
we are not finished since no mac yet... but linux is there now in the pre-release :)))))))

**Comment by Novantric** [comment]
Would it be possible to automatically detect an installed Dolphin Flatpak? Flatpaks don't have an executable aside from the what's exposed in `/var/lib/flatpak/exports/bin/org.DolphinEmu.dolphin-emu`.

**Comment by dominichayesferen** [comment]
> Would it be possible to automatically detect an installed Dolphin Flatpak? Flatpaks don't have an executable aside from the what's exposed in `/var/lib/flatpak/exports/bin/org.DolphinEmu.dolphin-emu`. It seems this is also causing an error where Dolphin can't open any files from `~/.config/CT-MKWII` or the ROM .iso's directory without creating an exception in Flatseal. Also, the License settings doesn't seem to work for me.

At worst, you could use the strat Warehouse uses to get installed Flatpaks - the 'run command outside of sandbox' ("arbitrary commands") permission.

**Comment by 9thyear2** [comment]
> Would it be possible to automatically detect an installed Dolphin Flatpak? Flatpaks don't have an executable aside from the what's exposed in `/var/lib/flatpak/exports/bin/org.DolphinEmu.dolphin-emu`. It seems this is also causing an error where Dolphin can't open any files from `~/.config/CT-MKWII` or the ROM .iso's directory without creating an exception in Flatseal. Also, the License settings doesn't seem to work for me.

a flatpak release might be a workaround for this. that way dolphin could be packaged as a dependency. but this would add another packaging format, and might require code changes as well... all for a workaround. and also if you had a non flatpak dolphin installed and then installed a flatpak of wheelwizard, it would download the flatpak version of dolphin and utilize that instead of using the already installed dolphin. now you have 2 dolphins on your system...

the solution that obs used for this is only ship officially as a flatpak (and also an ubuntu apt), and supports nothing else, everything else is unofficial. the reason they choose this is because flatpak allows them to control the dependencies, so when an issue occurs they know what version of what dependency you use. as for why they chose ubuntu apt as a second official method, i have no idea...

**but this is a pre-release**, for such a release it is most likely out of scope. especially since its the first build for linux, and thats only half the equation of this issue. as getting macOS support still needs to be completed...

so the current solution is kinda one where its good enough for now, but i wouldn't be surprised if the devs chose to package it differently in the future.

welcome to linux, the home of the great packaging war...

**Comment by Novantric** [comment]
Perhaps an appimage is in order, with some scripts to detect if Dolphin is installed as a Flatpak or via the system packages? This would also alleviate the need for WheelWizard to create the load directory.

**Comment by dominichayesferen** [comment]
So I've been told that a Flatpak can just... access another Flatpak's configs directory via filesystem permissions, so... theoretically all one would need to do for detecting Flatpak Dolphin in a Flatpak is just... have the permission to and check for the Flatpak Dolphin configs directory in order to append stuff to.

Similarly, obviously a Flatpak can access the non-Flatpak config directories via filesystem permissions as well, so that's easy enough once the Flatpak Dolphin portion is dealt with.

**Comment by gdmagana** [comment]
hey ive been testing on macos, i know you haven't done all the work to make it ready but i wanted to share a crash report from when the avalonia app was in the background while i played RR.

[crashlogAnonymized.txt](https://github.com/user-attachments/files/18961860/crashlogAnonymized.txt)

**Comment by patchzyy** [comment]
@gdmagana has made a mac version release you guys can check out in there
https://github.com/TeamWheelWizard/WheelWizardMac/releases/tag/v0.9.0-beta

it will be included in our upcoming releases aswell

so we will have support for
windows
windows-arm (untested)
mac-x64
mac-Arm
Linux-x64
Linux-Arm

**Comment by patchzyy** [comment]
https://github.com/TeamWheelWizard/WheelWizard/releases/tag/2.0.1

its out, let me know if there are any issues otherwise i will close this issue

**Comment by rehtlaw** [comment]
dunno if this is necessarily a Linux issue, but the window that automatically detects the flatpak userdir gets cut off

![Image](https://github.com/user-attachments/assets/2c6d6855-25b2-4867-9fd3-96376ad1d0d7)

**Comment by patchzyy** [comment]
noted, will work on it for the next minor update

**Comment by Octelly** [comment]
For anyone using Nix/NixOS: I got the launcher building and running here: [perma link to build file](https://github.com/OwO-Digital/nixos-config/blob/1d4e2c28cde2a42fcc7aefa93c21828cd323a68c/packages/games/WheelWizard/default.nix)

You can run it from your command line on any Linux system with Nix installed via ``nix run github:OwO-Digital/nixos-config/snowfall#WheelWizard``.

Any inputs welcome, not a .NET witch.

**Comment by Octelly** [comment]
Welp, seems like Flatpak would work, but native Doplhin doesn't start on NixOS, due to the hardcorded path to ``/bin/bash``. NixOS not being FHS-compliant cannot really supply this (this is my understanding, correct me if I'm wrong).

![Image](https://github.com/user-attachments/assets/199bfadf-3159-4fe0-894d-82ce18181fa9)

``/usr/bin/env bash`` seems to work and should work on most Linux systems in general (I recall using this in my scripts on Arch as well). Using this to invoke the shell both [here](https://github.com/TeamWheelWizard/WheelWizard/blob/df822c56fc15d23898ceb9b6599f432ae95f2961/WheelWizard/Services/Launcher/Helpers/DolphinLaunchHelper.cs#L36-L41) and in script shebangs should make the launcher work on NixOS as well.

**Comment by Octelly** [comment]
Sorry for spam, but I confirmed this being the issue by running the executable in the FHS environment NixOS uses to run Steam:

![Image](https://github.com/user-attachments/assets/e672128b-894e-436d-bb7e-486b27dff8e4)

**Comment by Octelly** [comment]
One more thing I ran into: I have RetroAchievements enabled by default, which made the mod only load partially. Took me a while to figure out what was causing the issue and I unfortunately did not find a command line option to disable RA temporarily for the session, but probably something worth figuring out.

**Comment by Octelly** [comment]
One last thing would be that the profiles thing isn't working for me. Just played a couple online races, but the profiles just aren't recognised by Wheel Wizard.

![Image](https://github.com/user-attachments/assets/0cdd42b8-2111-4069-8894-bca5972f8ad9)
